dpcpp/solver/batch_cg_kernels.hpp.inc

```diff
@@ -0,0 +1,244 @@
+    const ValueType& rho_old_shared_entry,
+    const ValueType* const __restrict__ z_shared_entry,
+    ValueType* const __restrict__ p_shared_entry, sycl::nd_item<3> item_ct1)
+{
+    const ValueType beta = rho_new_shared_entry / rho_old_shared_entry;
+    for (int li = item_ct1.get_local_linear_id(); li < num_rows;
+         li += item_ct1.get_local_range().size()) {
+        p_shared_entry[li] = z_shared_entry[li] + beta * p_shared_entry[li];
+    }
+}
+
+template <typename ValueType>
+__dpct_inline__ void update_x_and_r(
+    const int num_rows, const ValueType rho_old_shared_entry,
+    const ValueType* const __restrict__ p_shared_entry,
+    const ValueType* const __restrict__ Ap_shared_entry,
+    ValueType& alpha_shared_entry, ValueType* const __restrict__ x_shared_entry,
+    ValueType* const __restrict__ r_shared_entry, sycl::nd_item<3> item_ct1)
+{
+    auto sg = item_ct1.get_sub_group();
+    const auto tid = item_ct1.get_local_linear_id();
+    if (sg.get_group_id() == 0) {
+        single_rhs_compute_conj_dot_sg(num_rows, p_shared_entry,
+                                       Ap_shared_entry, alpha_shared_entry,
+                                       item_ct1);
+    }
+    item_ct1.barrier(sycl::access::fence_space::global_and_local);
+    if (tid == 0) {
+        alpha_shared_entry = rho_old_shared_entry / alpha_shared_entry;
+    }
+    item_ct1.barrier(sycl::access::fence_space::global_and_local);
+
+    for (int li = item_ct1.get_local_linear_id(); li < num_rows;
+         li += item_ct1.get_local_range().size()) {
+        x_shared_entry[li] += alpha_shared_entry * p_shared_entry[li];
+        r_shared_entry[li] -= alpha_shared_entry * Ap_shared_entry[li];
+    }
+    item_ct1.barrier(sycl::access::fence_space::global_and_local);
+}
+
+
+template <typename StopType, const int n_shared_total, typename PrecType,
+          typename LogType, typename BatchMatrixType, typename ValueType>
+__dpct_inline__ void apply_kernel(
+    const gko::kernels::batch_cg::storage_config sconf, const int max_iter,
+    const gko::remove_complex<ValueType> tol, LogType logger,
+    PrecType prec_shared, const BatchMatrixType& mat_global_entry,
+    const ValueType* const __restrict__ b_global_entry,
+    ValueType* const __restrict__ x_global_entry, const size_type num_rows,
+    const size_type nnz, ValueType* const __restrict__ slm_values,
+    sycl::nd_item<3> item_ct1,
+    ValueType* const __restrict__ workspace = nullptr)
+{
+    using real_type = typename gko::remove_complex<ValueType>;
+
+    const auto sg = item_ct1.get_sub_group();
+    const int sg_id = sg.get_group_id();
+    const int sg_size = sg.get_local_range().size();
+    const int num_sg = sg.get_group_range().size();
+
+    const auto group = item_ct1.get_group();
+    const auto batch_id = item_ct1.get_group_linear_id();
+
+    // The whole workgroup have the same values for these variables, but
+    // these variables are stored in reg. mem, not on SLM
```

Comment:
will it be only one shared value of whole workgroup?

Reply:
for each of them, yes.